--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -1920,15 +1920,15 @@ static void wpa_supplicant_event_assoc(s
 			wpa_supplicant_set_wpa_none_key(wpa_s,
 							wpa_s->current_ssid);
 		}
 		wpa_supplicant_cancel_auth_timeout(wpa_s);
 		wpa_supplicant_set_state(wpa_s, WPA_COMPLETED);
 	} else if (!ft_completed) {
 		/* Timeout for receiving the first EAPOL packet */
-		wpa_supplicant_req_auth_timeout(wpa_s, 10, 0);
+		wpa_supplicant_req_auth_timeout(wpa_s, 240, 0);
 	}
 	wpa_supplicant_cancel_scan(wpa_s);
 
 	if ((wpa_s->drv_flags & WPA_DRIVER_FLAGS_4WAY_HANDSHAKE) &&
 	    wpa_key_mgmt_wpa_psk(wpa_s->key_mgmt)) {
 		/*
 		 * We are done; the driver will take care of RSN 4-way
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -2607,15 +2607,15 @@ void wpa_supplicant_rx_eapol(void *ctx,
 	     wpa_s->current_ssid->mode != IEEE80211_MODE_IBSS)) {
 		/* Timeout for completing IEEE 802.1X and WPA authentication */
 		wpa_supplicant_req_auth_timeout(
 			wpa_s,
 			(wpa_key_mgmt_wpa_ieee8021x(wpa_s->key_mgmt) ||
 			 wpa_s->key_mgmt == WPA_KEY_MGMT_IEEE8021X_NO_WPA ||
 			 wpa_s->key_mgmt == WPA_KEY_MGMT_WPS) ?
-			70 : 10, 0);
+			70 : 60, 0);
 	}
 	wpa_s->eapol_received++;
 
 	if (wpa_s->countermeasures) {
 		wpa_msg(wpa_s, MSG_INFO, "WPA: Countermeasures - dropped "
 			"EAPOL packet");
 		return;
