--- a/scripts/kconfig/mconf.c
+++ b/scripts/kconfig/mconf.c
@@ -23,14 +23,18 @@
 #include <stdlib.h>
 #include <string.h>
 #include <strings.h> /* for strcasecmp */
 #include <termios.h>
 #include <unistd.h>
 #include <locale.h>
 
+#ifndef SIGWINCH
+#define SIGWINCH 28
+#endif
+
 #define LKC_DIRECT_LINK
 #include "lkc.h"
 
 static char menu_backtitle[128];
 static const char mconf_readme[] = N_(
 "Overview\n"
 "--------\n"
