--- a/util-linux/umount.c
+++ b/util-linux/umount.c
@@ -56,14 +56,17 @@ static struct mntent *getmntent_r(FILE*
 #define OPT_FORCE               (1 << 0) // Same as MNT_FORCE
 #define OPT_LAZY                (1 << 1) // Same as MNT_DETACH
 //#define OPT_FREE_LOOP           (1 << 2) // -d is assumed always present
 #define OPT_DONT_FREE_LOOP      (1 << 3)
 #define OPT_NO_MTAB             (1 << 4)
 #define OPT_REMOUNT             (1 << 5)
 #define OPT_ALL                 (ENABLE_FEATURE_UMOUNT_ALL ? (1 << 6) : 0)
+#ifndef MNT_DETACH
+#define MNT_DETACH 0x00000002
+#endif
 
 int umount_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int umount_main(int argc UNUSED_PARAM, char **argv)
 {
 	int doForce;
 	struct mntent me;
 	FILE *fp;
