Index: mac-telnet-2016-10-11/src/mactelnetd.c
===================================================================
--- mac-telnet-2016-10-11.orig/src/mactelnetd.c
+++ mac-telnet-2016-10-11/src/mactelnetd.c
@@ -448,6 +448,9 @@ static void user_login(struct mt_connect
 	}
 
 	if (user == NULL || memcmp(md5sum, curconn->trypassword, 17) != 0) {
+                /* Delay response to prevent brute force */
+                sleep(1);
+
 		syslog(LOG_NOTICE, _("(%d) Invalid login by %s."), curconn->seskey, curconn->username);
 
 		/*_ Please include both \r and \n in translation, this is needed for the terminal emulator. */
