--- a/iwconfig.c
+++ b/iwconfig.c
@@ -381,14 +381,22 @@ display_info(struct wireless_info *	info
 	}
     }
 
   /* Formating */
   if(tokens > 0)
     printf("\n          ");
 
+  /* display center frequency */
+  if(info->b.center1_freq)
+    {
+      iw_print_freq_value(buffer, sizeof(buffer), info->b.center1_freq);
+      printf("Center1-Freq: %s", buffer);
+      printf("\n          ");
+	}
+
   /* Display encryption information */
   /* Note : we display only the "current" key, use iwlist to list all keys */
   if(info->b.has_key)
     {
       printf("Encryption key:");
       if((info->b.key_flags & IW_ENCODE_DISABLED) || (info->b.key_size == 0))
 	printf("off");
@@ -802,14 +810,56 @@ set_freq_info(int		skfd,
 
   /* Var args */
   return(i);
 }
 
 /*------------------------------------------------------------------*/
 /*
+ * Set center1 frequency
+ */
+static int
+set_center1_info(int		skfd,
+	      char *		ifname,
+	      char *		args[],		/* Command line args */
+	      int		count)		/* Args count */
+{
+	struct iwreq		wrq;
+	int			i = 1;
+	double		freq;
+	char *		unit;
+
+	if(!strcasecmp(args[0], "auto")) {
+		wrq.u.freq.m = -1;
+		wrq.u.freq.e = 0;
+	} else {
+		freq = strtod(args[0], &unit);
+		if(unit == args[0]) {
+			errarg = 0;
+			return(IWERR_ARG_TYPE);
+		}
+		if(unit != NULL) {
+			if(unit[0] == 'G') freq *= GIGA;
+			if(unit[0] == 'M') freq *= MEGA;
+			if(unit[0] == 'k') freq *= KILO;
+		}
+
+		iw_float2freq(freq, &(wrq.u.freq));
+	}
+
+	wrq.u.freq.flags = 0xc1;
+
+	if(iw_set_ext(skfd, ifname, SIOCSIWFREQ, &wrq) < 0)
+		return(IWERR_SET_EXT);
+
+	/* Var args */
+	return(i);
+}
+
+/*------------------------------------------------------------------*/
+/*
  * Set Bit Rate
  */
 static int
 set_bitrate_info(int		skfd,
 		 char *		ifname,
 		 char *		args[],		/* Command line args */
 		 int		count)		/* Args count */
@@ -1704,14 +1754,16 @@ static const struct iwconfig_entry iwcon
 	"Set ESSID",			"{NNN|any|on|off}" },
   { "mode",		set_mode_info,		1,	SIOCSIWMODE,
 	"Set Mode",			"{managed|ad-hoc|master|...}" },
   { "freq",		set_freq_info,		1,	SIOCSIWFREQ,
 	"Set Frequency",		"N.NNN[k|M|G]" },
   { "channel",		set_freq_info,		1,	SIOCSIWFREQ,
 	"Set Frequency",		"N" },
+  { "center1",		set_center1_info,		1,	SIOCSIWFREQ,
+	"Set Center1 Frequency",		"N.NNN[k|M|G]" },
   { "bit",		set_bitrate_info,	1,	SIOCSIWRATE,
 	"Set Bit Rate",			"{N[k|M|G]|auto|fixed}" },
   { "rate",		set_bitrate_info,	1,	SIOCSIWRATE,
 	"Set Bit Rate",			"{N[k|M|G]|auto|fixed}" },
   { "enc",		set_enc_info,		1,	SIOCSIWENCODE,
 	"Set Encode",			"{NNNN-NNNN|off}" },
   { "key",		set_enc_info,		1,	SIOCSIWENCODE,
--- a/iwlib.c
+++ b/iwlib.c
@@ -698,14 +698,20 @@ iw_get_basic_config(int			skfd,
   /* Get frequency / channel */
   if(iw_get_ext(skfd, ifname, SIOCGIWFREQ, &wrq) >= 0)
     {
       info->has_freq = 1;
       info->freq = iw_freq2float(&(wrq.u.freq));
       info->freq_flags = wrq.u.freq.flags;
     }
+  /* Get Center frequency (0xC1) */
+  wrq.u.freq.flags = 0xc1;
+  if(iw_get_ext(skfd, ifname, SIOCGIWFREQ, &wrq) >= 0)
+    {
+      info->center1_freq = iw_freq2float(&(wrq.u.freq));
+    }
 
   /* Get encryption information */
   wrq.u.data.pointer = (caddr_t) info->key;
   wrq.u.data.length = IW_ENCODING_TOKEN_MAX;
   wrq.u.data.flags = 0;
   if(iw_get_ext(skfd, ifname, SIOCGIWENCODE, &wrq) >= 0)
     {
--- a/iwlib.h
+++ b/iwlib.h
@@ -173,14 +173,15 @@ typedef struct wireless_config
 {
   char		name[IFNAMSIZ + 1];	/* Wireless/protocol name */
   int		has_nwid;
   iwparam	nwid;			/* Network ID */
   int		has_freq;
   double	freq;			/* Frequency/channel */
   int		freq_flags;
+  double	center1_freq;			/* Center Frequency */
   int		has_key;
   unsigned char	key[IW_ENCODING_TOKEN_MAX];	/* Encoding key used */
   int		key_size;		/* Number of bytes */
   int		key_flags;		/* Various flags */
   int		has_essid;
   int		essid_on;
   char		essid[IW_ESSID_MAX_SIZE + 1];	/* ESSID (extended network) */
